src/org/sosy_lab/java_smt/solvers/cvc5/CVC5InterpolatingProver.java

```diff
@@ -0,0 +1,370 @@
+    ImmutableList<Collection<Term>> formBAsList = ImmutableList.of(formulasOfB);
+
+    Term itp = getCVC5Interpolation(formAAsList, formBAsList);
+
+    return creator.encapsulateBoolean(itp);
+  }
+
+  @Override
+  public List<BooleanFormula> getSeqInterpolants(
+      List<? extends Collection<Term>> partitionedFormulas)
+      throws SolverException, InterruptedException {
+    Preconditions.checkArgument(
+        !partitionedFormulas.isEmpty(), "at least one partition should be available.");
+
+    List<BooleanFormula> itps = Stream.of(new BooleanFormula[] {}).collect(Collectors.toList());
+    for (int i = 1; i < partitionedFormulas.size(); i++) {
+      itps.add(
+          getInterpolant(
+              ImmutableList.copyOf(Iterables.concat(partitionedFormulas.subList(0, i)))));
+    }
+    return itps;
+  }
+
+  // Experimental!
+  @Override
+  public List<BooleanFormula> getTreeInterpolants(
+      List<? extends Collection<Term>> pPartitionedFormulas, int[] pStartOfSubTree)
+      throws SolverException, InterruptedException {
+    Preconditions.checkState(!closed);
+    assert InterpolatingProverEnvironment.checkTreeStructure(
+        pPartitionedFormulas.size(), pStartOfSubTree);
+
+    // Generate every Interpolation Pair
+    ImmutableList<ImmutableList<ImmutableList<Collection<Term>>>> interpolationPairs =
+        getTreeInterpolationPairs(pPartitionedFormulas, pStartOfSubTree);
+
+    final List<Term> itps = Stream.of(new Term[] {}).collect(Collectors.toList());
+    try { // Interpolate every Interpolation Pair
+      for (int i = 0; i < interpolationPairs.size(); i++) {
+        itps.add(
+            getCVC5Interpolation(
+                interpolationPairs.get(i).get(0), interpolationPairs.get(i).get(1)));
+      }
+    } catch (UnsupportedOperationException e) {
+      if (e.getMessage() != null) {
+        throw new SolverException(e.getMessage(), e);
+      } else {
+        throw e;
+      }
+    }
+
+    final List<BooleanFormula> result =
+        Stream.of(new BooleanFormula[] {}).collect(Collectors.toList());
+    for (Term itp : itps) {
+      result.add(creator.encapsulateBoolean(itp));
+    }
+    assert result.size() == pStartOfSubTree.length - 1;
+    return result;
+  }
+
+  /**
+   * Interpolates a Tuple of Interpolants according to Craig-Interpolation using CVC5-Interpolation.
+   *
+   * <p>CVC5's getInterpolant: There is a Model, such that the Interpolant I, with A -> I = True
+   * (1CVC5) and I -> B = True (2CVC5).
+   *
+   * <p>Craig Interpolation: There is a Model, such that the Interpolant psi, with phi- -> psi =
+   * True (1Craig), not(psi /\ phi+) = True (2Craig).
+   *
+   * <p>With A/phi- current set of assertions and B/phi+ Formulas to interpolate.
+   *
+   * <p>CVC5 -> Craig Interpolation:
+   *
+   * <p>(1CVC5) <=> (1Craig) due to subst of A with phi- and reflexivity.
+   *
+   * <p>(2CVC5) <=> I -> B = True <=> (not I) \/ B = True <=> not (I /\ (not B)) = True <=> (2Craig)
```

Comment:
Are you sure, you can convert SAT to UNSAT in that case? I am not yet convinced.

Reply:
Which part seems problematic? This one: 
```
not (I /\ (not B)) = True <=> I /\ (not B) = False <=> (2Craig)
```
?

Reply:
Your "proof" only transforms the formula and ignores the fact, that the query for the solver changes from "sat" to "unsat". This just feels weird..., I currently have no counterexample, why it should not work. Perhaps it is valid overall. :-)

Maybe discuss with your supervisor or in a SosyLab group meeting for details.

Reply:
I did not meet with my supervisor yet, but I want to discuss: 
 - `1CVC5 <=> 1Craig`: both formulas need to hold/be valid, so there should be no problem.
 - `2CVC5 <=> 2Craig`: is the swap from valid to "unsat" not represented by `...=True -> ...=False`? In CVC5 it is always valid (for me that means equals `True`, different to "sat") in Craig it is always "unsat" (the formula equals `False`). So there happens a swap by just transforming formulas, or am I mixing things up?

Reply:
I added a precondition to check that CVC5 satisfies the definition of Craig-Interpolation. It held for every test case.